오픈마켓 II [STEP 1] mmim, cathy

JoSH0318 opened the pull request:
@Monsteel 토니!!😭
pr이 너무 늦어 죄송합니다.
정말 생각보다 더 어렵네요...
잘부탁드립니다 😎

---

## II STEP 1
### 고민한 점(트러블 슈팅)
> 0️⃣ 이번 프로젝트는 무지함에서 오는 고민이 많았다. 여러가지 선택지나 가설을 통한 고민이 아닌 전혀 모르는 영역을 탐구하는 고민이 대부분이다. 

1️⃣ 상품 등록/수정 페이지 UI 구현
- 상품 등록/수정 페이지는 크게 3 파트로 나뉜다.
    - 상단: 이미지가 들어가는 횡스크롤
    - 중단: 텍스트필드가 들어있는 입력란들
    - 하단: 긴 줄글을 입력하는 텍스트뷰
- 이번 프로젝트의 가장 큰 목표 중에 CollectionView의 구현이 있다.
- 따라서 CollectionView를 통한 구현을 고민했다.
- 하지만 몇 가지 이유로 적합하지 않다고 판단됐다.
    1. 굳이 Cell을 사용한 화면구현이 필요없다. CollectionView, TableView가 사용되는 가장 큰 이유는 반복되는 Cell들을 사용하기 위한이라고 생각한다. 이 화면에서는 반복되는 Cell이 없다고 판단됐다.
    2. Cell protoType이 4가지 이상 존재해야한다.(이미지뷰, 버튼이 있는 뷰, 텍스트필드뷰, 텍스트뷰 등이 구현된 Cell) 하지만 이 요소들이 반복되지 않고, 한번만 쓰인다.
    3. 상단의 scrollView 영역만 collectionView를 쓰는것도 생각해봤다. 하지만 이미지는 5개 이상 생성되지 않는다. 따라서 많은 Cell을 반복적으로 사용하는 collectionView에 적합하지 않다.
> 따라서 상단은 scrollView, 중단은 stackView, 하단은 TextView로 구현하기로 결정했다.  
![](https://i.imgur.com/sXgYrv9.png)

2️⃣ HTTP POST 구현
- 일단 POST API 명세서를 보면 헤더와 바디에 필요한 요구사항들을 확인할 수 있다.
- 헤더 정보는 `request.setValue`를 통해 간단히 명시할 수 있다.
- 하지만 `params`와 `images`가 다른 타입을 가졌기에 `Multipart/form-data`을 사용해야했다.
- 일단 Multipart/form-data에 대한 고민을 했다.
- Multipart/form-data는 다양한 타입을 POST할 때 사용하는 방식이다.
- 때문에 헤더에 Content-Type으로 Multipart/form-data을 명시하고, boundary를 통해 다른 두 타입을 가진 각각의 바디를 구분하여 생성해주었다.
```
----------
< header >
* 필요한 header정보 입력
* Content-Type: Multipart/form-data
----------
< Body >
* Content-Disposition: MIME Type
* Content-Type: 해당 바디의 Type
* data

----------
< Body >
* Content-Disposition: MIME Type
* Content-Type: 해당 바디의 Type
* data

----------
```
- 서버가 요구하는 명확한 요구조건과 HTTP을 준수해야하기 때문에 오타가 날 경우, 통신이 안된다.
- 오타와 같은 경우 그것을 찾기가 매우 어렵고, 많은 시간이 소모되는 것을 경험했다.
- 때문에 오타와 문법에 각별히 신경써야했다.

3️⃣ image POST
- Multipart/form-data을 통해 image를 POST하기 위한 Body를 구현할 때 고민이 많았다.
- 로컬에 저장된 image의 경로 정보를 통해 image를 POST하거나, image의 URL 정보를 POST하는게 아닐까 생각했다. 이런 잘못된 예상으로 인해 많은 시간을 소모했다.
- API 명세를 살펴보면 image의 타입은 `file[]`로 되어있다. 
- 때문에 URL과 같은 String타입 POST하는 것이 아니라는 새로운 가설을 세웠고, data 형식으로 http body에 append해주는 방법을 생각하게 됐다.
- imagePickerController를 통해 선택된 사진을 UIImgaeView에 뿌리고, 그 UIImageView를 data형식으로 변환하여 http body에 append해주었다.

4️⃣ 키보드가 화면을 가리는 문제
- `textView` 입력할 때 키보드에 화면이 가려지는 문제가 있었다. 
- `Notification`을 사용해서 뷰의 frame을 키보드 높이의 1/3만큼 올려주는 방법으로 해결하려했다. 
- 또한 `UIPanGestureRecognizer`를 사용해서 사용자가 아래로 스와이프했을 때 키보드를 내려주도록 구현하였다. 
- `UITextViewDelegate`를 통해 textView에서만 적용되게 구현하려했지만 textView가 아닌 다른 textField에도 모두 키보드에 대한 설정이 적용되어서 결국 해결하지 못했다.

5️⃣ 모달로 화면 전환하는 방식
```swift
    let registrationVC = RegistrationViewController()
    let navigationController = UINavigationController(rootViewController: registrationVC)
    navigationController.modalPresentationStyle = .fullScreen
    present(navigationController, animated: true, completion: nil)
```
위와 같이 매개변수로 루트뷰 컨트롤러가 될 뷰컨트롤러를 생성하고 그 뷰컨트롤러로 네비게이션 컨트롤러를 생성해주었다.

6️⃣ POST와 비동기
- Post를 하기 위해 navigationItem Done button을 트리거로 이용했다.
- button을 클릭하면 아래와 같은 코드가 실행되게 구현했다.
```swift
self.apiProvider.patch(.editing(productId: product.id), params) { result in
    switch result {
    case .success(_):
        return
    case .failure(let response):
        return
    }
}
self.dismiss(animated: true, completion: nil)
```
- 하지만 이렇게 화면을 빠져나가 상품 리스트 화면으로 가게되면, post한 상품이 반영되지 않는다.
- 이러한 버그가 생기는 이유에 대해서 몇가지 고민했고, 이러한 방법을 통해 해결했다.
    1. post(서버와 통신)하는 메서드는 비동기이다. 때문에 switch문부터 시작하는 내부 코드블럭을 실행하기 이전에 dismiss가 된다. 따라서 화면을 빠져나간 시점에는 아직 post가 완료되지 않아 갱신되지 않는다. (post보다 상품리스트의 get을 통한 갱신이 더 빠르다.)
    이 문제를 해결하기 위해 DispatchGroup을 통해 patch 메서드 내부 코드 블럭이 끝날 때까지 기다려주도록 했다. group.notify을 사용했고, 후행 클로저에 dismiss하는 코드를 구현했다.
    3. 상품 리스트 화면의 viewWillapear 시점에 get을 통해 상품을 갱신한다. get 또한 비동기 메서드 이기 때문에 get을 완료하기 이전에 이미 viewWillApear 시점이 끝나버린다. (현재 가진 정보로 화면을 구성해버히고 지난간 느낌) 따라서 viewDidApear 시점에 get 메서드를 사용했다. 이미 화면의 라이프 사이클은 시작됐고, 그 직후 갱신된 정보를 반영하는 식으로 해결할 수 있었다. 

7️⃣ 빈 문자열을 get, post할 수 있다?
- 현재 서버는 빈 문자열 POST을 허용한다. price와 같은 Int 타입도 `" "`빈 문자열을 받을 수 있고, 예상하기론 nil값으로 서버에 저장되는 것 같다.
- 따라서 정보중 하나라도 nil 상태의 상품을 Get하게 된다면, 디코드 오류가 생기게 된다. 
- 현재 step까지는 상품 상세 화면을 구현하지 않아도 되기 때문에 이러한 nil 정보에 대해서 신경쓰지 않아도 된다고 판단했다.
- 하지만 이번 프로젝트에서 POST나 PATCH 과정에서 빈 문자열 또는 잘못된 문자열을 보낼수 없도록 막는 것이 좋다고 판단했다.
- name, price, discounted_price, stock, description 중 하나라도 입력이 누락됐다면 경고 Alert을 띄우는 것이 사용자 경험상 좋을 것이라고 생각했다. (📎입력을 누락하게 되면 빈문자열("")이 입력된다.) 
- 또한 숫자를 입력해야하는 곳에 Int로 변환할 수 없는 값을 입력하면 0 값이 입력되게 구현했다. 
> 이 후에 정보를 입력하지 않은 textField의 하이라이트를 주는 등 사용자 경험 측면에서 좀더 업그레이드 하고 싶다.


### 질문사항
1️⃣ textview부분만 키보드 내리는 것 적용하는 방법
프로젝트 요구사항을 보면 키보드가 올라왔을 때 컨텐츠를 가리지 않도록 한다는 조건이 있습니다. 
![](https://i.imgur.com/amUTGrJ.png)  
중간에 있는 **상품명 ~ 재고수량 TextField**는 키보드가 가릴 일이 없습니다. 하지만 아래에 있는 **description textView**와 같은 경우는 가립니다!
따라서 **description textView**를 클릭하여 키보드가 올라왔을 때만 view.origin.y를 올려주려했습니다 😭
하지만 `NotificationCenter`가 키보드가 올라오는 액션(UIResponder.keyboardWillShowNotification)를 감지하여 origin을 올려주기 때문에  **description textView**의 키보드가 올라왔을 때만 origin을 올려주지 못했습니다.
그래서 현재는 모든 키보드가 올라올 때, view.origin.y를 올라갑니다 😭
어떠한 키워드로 공부하면 이점을 해결할 수 있을지, 또는 어떤 문제가 있는지 궁금합니다.🙇‍♂️


2️⃣ 공통부분 상속으로 구현했을 때 private 접근제어자 사용하지 못하는 문제
RegistrationView와 EditingView가 완벽히 겹치는 코드가 많아서 프로토콜과 상속을 통한 구현을 시도해봤습니다. 
우선 protocol을 통해 구현하려고 했습니다. 저희 코드는 저장 프로퍼티를 많이 사용하고, 그 프로퍼티가 완벽히 겹칩니다. 때문에 protocol을 통해 구현하면 완벽히 같은 코드블럭도 똑같이 작성해야했습니다.(물론 코드 누락을 피할 수 있다는 장점이 생겼지만 중복코드가 여전히 존재하는 단점이...) 
그래서 현재 ProductUpdator class를 통해 상속으로 구현했습니다. 하지만 super class의 모든 프로퍼티와 메서드의 접근제어자를 사용할 수 없어졌습니다. 결과적으로 은닉화/캡슐화가 잘된 것인가 라는 생각을 하게 됐습니다. 
이 부분에서 저희가 사용한 상속이 올바른 방법인지 다른 방법이 있는지 궁금합니다.❤️‍🔥

3️⃣ 빈 문자열이 post 가능한 것에 대한 질문
위의 고민한 점에서(7️⃣ 빈 문자열을 get, post할 수 있다?) 언급했듯이 httpGet 했을때, price와 같은 Int 타입의 값에서 nil 값이 나옵니다! 
예상하기론 
1. 서버에서 빈 값을 nil로 바꾸어 저장했다. 
2. 서버에서 Int 타입임에도 ""로 가지고 있다가 클라이언트로 넘어오면서 Int타입으로 받을 수 없어서 nil을 반환한다.
이렇게 두가지로 예상해봤습니다.
궁금한 점은 POST할 때 params의 price는 type이 number라고 기재되어있는데, 왜 빈배열("")을 POST 할 수 있는건지. 서버에서 어떤 방식으로 잘못된 정보를 취급하는 것인지 궁금합니다.🙇‍♀️

4️⃣ 이외에도 너무 어려운 것이 많아서 아직 질문조차 못 드리는 것이 많은 것 같아요. 잘 부탁드립니다 😭

JoSH0318 committed:
✨[feat]: 상품등록 view인 RegistrationView 구현

*RegistrationViewController 파일 생성 및 초기 구현

Commit:
✨[feat]: RegistrationViewController로 화면 전환하는 기능 구현

Commit:
✨[feat]: RegistrationViewController 구현

Commit:
♻️[refactor]: RegistrationView 수정

JoSH0318 committed:
✨[feat]: inof.plist, 로컬에서 사진가져오는 property list 추가

JoSH0318 committed:
🎨[style]: 필요없는 개행 제거

JoSH0318 committed:
🪛[chore]: private 키워드 제거

JoSH0318 committed:
🪛[chore]: 잘못된 textField 출력문구 변경

JoSH0318 committed:
🪛[chore]: private 키워드 제거

JoSH0318 committed:
🎨[style]: 팀 컨벤션에 따른 코드 수정

JoSH0318 committed:
✨[feat]: UIImagePickerController를 이용한 사진 불러오기 기능 구현

Commit:
✨[feat]: Params, ImageFile struct 구현

Commit:
♻️[refactor]: 이미지 crop 기능 추가

Commit:
♻️[refactor]: Endpoint case 추가

Commit:
♻️[refactor]: textField keyboard 기능 추가 및 레이아웃 변경

Commit:
✨[feat]: http post 기능 초기 구현

JoSH0318 committed:
✨[feat]: extension Data 구현

JoSH0318 committed:
♻️[refactor]: reqest body 생성하는 로직 추가 구현

Commit:
🪛[chore]: 프로퍼티 네이밍 변경

Commit:
✨[feat]: setupParams 메서드 구현

Commit:
✨[feat]: postData, convertToImageFile 메서드 구현

JoSH0318 committed:
🐛[fix]: 오타로 인한 POST 버그 수정

JoSH0318 committed:
♻️[refactor]: postData() post 내부 로직 추가

JoSH0318 committed:
✨[feat]: image의 사이즈를 줄이는 resizeImage()메서드 구현

Commit:
✨[feat]: EditView 파일 생성

Commit:
♻️[refactor]: post 끝낸 후 productList로 나가는 기능 추가

JoSH0318 committed:
✨[feat]: EditView 삭제, 수정화면을 위한 EditingView 구현

JoSH0318 committed:
✨[feat]: 상세화면 정보를 디코드할 데이터 타입 DetailProduct 구현

JoSH0318 committed:
♻️[refactor]: 서버 통신을 위한 params secret,identifier 수정

JoSH0318 committed:
✨[feat]: 상세화면 통신을 위한 Endpoint 추가

JoSH0318 committed:
✨[feat]: 수정화면 editingView 프로퍼티 생성

JoSH0318 committed:
✨[feat]: collectionView Delegate 구현

Commit:
✨[feat]: EditingViewController 구현

configureRegistration, configureNavigationBar 메서드 생성

Commit:
✨[feat]: DetailViewController 구현

configureNavigationBar, presentEditingView 메서드 생성

Commit:
♻️[refactor]: detailViewController로 화면전환하는 로직 추가

JoSH0318 committed:
🚚[move]: swift 파일 위치 변경

JoSH0318 committed:
♻️[refactor]: Image, Vendor 구조체 타입 생성

JoSH0318 committed:
♻️[refactor]: DetailProduct 내부 프로퍼티 타입 변경

JoSH0318 committed:
♻️[refactor]: MainVC->DetailVC delegate로 데이터 주입

JoSH0318 committed:
✨[feat]: 해당 상품 정보를 위한 fetchDetailProductData() 구현

JoSH0318 committed:
✨[feat]: EditingVC에 상품정보를 뿌려주는 로직 구현

Commit:
♻️[refactor]: 이미지를 5장이상 추가 시 버튼 hidden 기능 구현

Commit:
✨[feat]: UIViewController extension 구현

Commit:
🪛[chore]: 개행 제거

Commit:
♻️[refactor]: descriptionTextView로 네이밍 변경

Commit:
♻️[refactor]: keyboard 관련 메서드 추가 및 팀 컨벤션 맞게 코드 수정

JoSH0318 committed:
🐛[fix]: MainVC로 다시 돌아왔을때, identifier 중복 문제 해결

*UUID로 디코드 데이터 식별자 변경

JoSH0318 committed:
🐛[fix]: POST가 완료되지 않은 시점에서 화면이 dismiss되는 문제 해결

JoSH0318 committed:
♻️[refactor]: self.fetchPages()메서드 호출 시점 변경

Commit:
✨[feat]: EditingProduct 타입 생성

Commit:
✨[feat]: http patch를 위한 기능 구현

Commit:
✨[feat]: 상품 수정을 위한 로직 구현

JoSH0318 committed:
✨[feat]: 상품 추가, 수정 view의 super view 생성 및 구현

JoSH0318 committed:
♻️[refactor]: ProductUpdaterView 상속에 따른 코드 수정

JoSH0318 committed:
🐛[fix]: 서버에서 일정 데이터가 nil 값이 올 경우 decode 실패하는 오류 해결

*nil 값이 존재할 가능성이 있는 값들을 optional로 처리

JoSH0318 committed:
♻️[refactor]: ImageFile 구조체 내부 프로퍼티에 기본값 할당

JoSH0318 committed:
✨[feat]: enum Constants 생성 및 그에 따른 로직 수정

JoSH0318 committed:
✨[feat]: enum Constants 생성 및 그에 따른 로직 수정

JoSH0318 committed:
✨[feat]: String을 Int 타입으로 변환해주는 연산 프로퍼티 구현

JoSH0318 committed:
♻️[refactor]: setupParams() 메서드 반환값 및 내부 로직 변경

JoSH0318 committed:
♻️[refactor]: name space  Constants 프로퍼티 변경 및 추가

JoSH0318 committed:
✨[feat]: 누락된 정보가 있을 시, 경고 Alert을 띄우는 기능 구현

JoSH0318 committed:
✨[feat]: 누락된 정보가 있을 시, 경고 Alert을 띄우는 기능 구현

Commit:
♻️[refactor]: displayInformation으로 메서드 이름 변경

Commit:
♻️[refactor]: EditingVC 네임스페이스 추가 및 메서드 위치 변경

Commit:
♻️[refactor]: DetailVC 내 메서드 위치 변경

Commit:
♻️[refactor]: RegistrationVC 프로퍼티 및 메서드 위치 변경

JoSH0318 committed:
♻️[refactor]: 화면간 데이터 통신 방법, 프로퍼티 주입에서 메서드 주입으로 변경

JoSH0318 committed:
🔥[del]: Mock을 이용한 테스트 방법 변경으로 쓸모없어진 코드 및 파일 삭제

JoSH0318 committed:
✅ [test]: APIProvider 제네릭 제거로 인한 테스트 코드 변경

JoSH0318 committed:
♻️[refactor]: ApiProvider 제네릭 제거 및 메서드 통합

JoSH0318 committed:
♻️[refactor]: ApiProvider 제네릭 제거로 인한 코드 변경

Commit:
📝[docs]: 오픈마켓 II STEP1 README update

Commit:
✨[feat]: Currency enum 구현

Commit:
♻️[refactor]: 매직리터럴 제거 및 if문을 switch으로 변경

Commit:
♻️[refactor]: ""을 isEmpty로 변경

JoSH0318 committed:
♻️[refactor]: enum Currency, rawValue String으로 변경

JoSH0318 committed:
♻️[refactor]: enum Currency 파일 위치 변경 및 var text 추가

JoSH0318 committed:
♻️[refactor]: enum Currency 변경으로 인한 코드 변경

JoSH0318 committed:
♻️[refactor]: CollectionView Pagination 기능 변경

JoSH0318 committed:
♻️[refactor]: setupParams() 메서드, verifyCurrency()로 기능 분리

Monsteel commented on OpenMarket/OpenMarket/Model/ApiProvider.swift:
```diff
@@ -9,26 +9,18 @@ import Foundation
 enum HttpMethod {
   static let get = "GET"
+  static let post = "POST"
+  static let patch = "PATCH"
```
하핫! 추가가 되었군요! ✨

Monsteel commented on OpenMarket/OpenMarket/Model/Currency.swift:
```diff
@@ -0,0 +1,26 @@
+    return self.rawValue
+  }
+  
+  var number: Int {
```
여기 이 달러와 원을 왜 number(0/1)로 표현하나요?! 변수명이 너무 포괄적인 의미다 보니 의도를 파악하기 어렵습니다 😵

JoSH0318 replied:
segmented Control의 특별한 option을 따른 수인데 그냥 number로 해서 의도 파악이 힘들것 같다는 생각을 했습니다!
optionNumber로 변경해보았습니다.

Monsteel commented on OpenMarket/OpenMarket/Model/EditingProduct.swift:
```diff
@@ -0,0 +1,25 @@
+  let stock: Int
+  let secret: String
+  
+  enum CodingKeys: String, CodingKey {
```
이곳의 CodingKeys도 private 접근제어자를 사용할 수 있을것 같습니다 🙋🏻‍♂️

Monsteel commented on OpenMarket/OpenMarket/Model/Params.swift:
```diff
@@ -0,0 +1,29 @@
+}
+
+struct ImageFile {
+  let fileName: String = "imageName.jpeg"
```
파일 이름을 `imageName.jpeg`로 고정하신 이유가 있으신가요?! 🙋🏻‍♂️

JoSH0318 replied:
사실  fileName이 존재하는 이유를 아직까지 모르겠습니다. 😿
저희가 fileName을 넣어준 이유는 API 명세서의 예시에 있어서 꼭 필요한 값이지 않을까라는 생각에 넣게 됐습니다.

어떠한 식별자의 역할도 아니고, 이미지를 다운 받을 때 다운받은 이미지의 이름이 fileName도 아닌데... 
저것이 왜 필요한지 궁금해요.🥺

그리고 만약 필요한 값이라면 어떠한 값을 넣어줘야하나요?
image picker를 통해 선택한 image의 이름? image의 경로? 랜덤한 이름? 어떤 값을 넣어야할지 모르겠어요 😭

Monsteel replied:
@JoSH0318 해당 이미지의 파일 이름을 서버에 알려주는 역할을 합니다.
이번 서버에서는 별도의 식별자를 부여해서 파일을 관리하여 다른 문제가 생기지 않는것으로 추측이 됩니다.

당장 문제는 발생하지 않지만, 요 부분은 서버의 구현에 의해 달라질 수 있는 만큼
파일 이름을 하나로 통일하는 것 보다, 해당 이미지를 식별할 수 있는 이름을 부여하거나, 원본 이미지의 파일이름을 정하여 서버에게 전송하는건 어떨까요?

대부분 날짜와 시간으로 중복되기 어려운 값들로 고유한 값을 정하거나, UUID와 같이 랜덤한 값을 부여하기도 합니다!

Monsteel commented on OpenMarket/OpenMarket/Util/Extensions/String+Extension.swift:
```diff
@@ -8,6 +8,14 @@
+    if let number = Int(self) {
+      return number
+    } else {
+      return 0
```
실제 숫자 0인지 오류에 의해 return된 0인지 명확하지 않아 뭔가 불안하네요 😥

JoSH0318 replied:
nil로 처리하고 text들의 조건을 검사하는 로직을 추가했습니다!

Monsteel commented on OpenMarket/OpenMarket/View/ProductUpdaterView.swift:
```diff
@@ -0,0 +1,179 @@
+
+import UIKit
+
+class ProductUpdaterView: UIView {
```
Base~ 를 사용하지 않은 ProductUpdaterView 네이밍 좋네요! 🙌

Monsteel commented on OpenMarket/OpenMarket/Model/Products.swift:
```diff
@@ -59,11 +60,11 @@ struct Page: Decodable, Hashable {
+    lhs.uuid == rhs.uuid
   }
 }
 
```
Model의 이름에 `Products`라는 복수형 s가 붙어있지만, 하나의 Product의 내용을 담고 있어, 복수형으로 표현하는건 옳지 않아 보입니다. ProductsList 라는 이름도 마찬가지 입니다. List의 element를 표현하고자 하신걸로 보이는데, 그 의도를 조금 더 잘 드러낼 수 있는 네이밍을 고민해보면 좋을것 같아요 ✍🏻

Monsteel commented on OpenMarket/OpenMarket/Model/ApiProvider.swift:
```diff
@@ -44,12 +36,92 @@ struct ApiProvider<T: Decodable> {
+    return body
+  }
 }
+
```
ApiProvider를 Model로 분류하신 이유가 있을까요?! 역할로 보면, Util에 가까운것 같아서요!
그리고, API는 Application Programming Interface 의 약자로, 밈과 캐시가 정한 그라운드룰의 [코딩 컨벤션](https://github.com/StyleShare/swift-style-guide#%EC%95%BD%EC%96%B4)에 어긋납니다! 😵

JoSH0318 replied:
일단 현재 ApiProvider라는 이름이 객체의 역할과 조금 어울리지 않아 HttpProvider로 변경했습니다.🙇‍♀️
저희 코딩 컨벤션에는 약어로 시작할 땐, 소문자로 시작한다고 되어있습니다.🙇‍♂️
그런데 property, method와 같은경우에만 해당된다고 생각했고, class, struct와 같은 경우는 대문자로 시작한다고 생각했습니다.🥺

사실, 구체적인 예시가 없어서 저희가 의논한 결과 class, struct와 같은 경우는 약어라고해도 맨앞은 대문자로 하기로 결정해봤습니다 !😀

Monsteel reviewed (changes requested):
안녕하세요 밈 캐시~🙌
Step1 수고 많으셨습니다! 

몇가지 코멘트와 질문주신 내용에 대해 답변 남겨보았습니다! 확인 부탁드려요 ☺️

> 1️⃣ textview부분만 키보드 내리는 것 적용하는 방법
> 프로젝트 요구사항을 보면 키보드가 올라왔을 때 컨텐츠를 가리지 않도록 한다는 조건이 있습니다. 
![](https://i.imgur.com/amUTGrJ.png) 
> 중간에 있는 **상품명 ~ 재고수량 TextField**는 키보드가 가릴 일이 없습니다. 하지만 아래에 있는 **description textView**와 같은 경우는 가립니다!
따라서 **description textView**를 클릭하여 키보드가 올라왔을 때만 view.origin.y를 올려주려했습니다 😭
하지만 `NotificationCenter`가 키보드가 올라오는 액션(UIResponder.keyboardWillShowNotification)를 감지하여 origin을 올려주기 때문에  **description textView**의 키보드가 올라왔을 때만 origin을 올려주지 못했습니다.
그래서 현재는 모든 키보드가 올라올 때, view.origin.y를 올라갑니다 😭
어떠한 키워드로 공부하면 이점을 해결할 수 있을지, 또는 어떤 문제가 있는지 궁금합니다.🙇‍♂️

밈과 캐시가 하고자 하는 부분을 구현하려면, 단순히 키보드의 높이만 빼주는 방식으로는 어려울 것 같아요.
얼마나 빼주어야 하는지, 뷰가 어디에 위치해야 하는 지 등을 판단할 수 있는 포커싱된 Testfield의 좌표가 필요할 것 같아요. 그리고 꼭 하나의 selectora만 addObserver 해주지 않아도 됩니다!

여러 방법으로 해당 문제를 해결할 수 있을것 같은데, 본질적인 해결책을 도출하고 싶으시다면,
사용중심의 학습보다는 조금 더 본질을 파고드는 방향으로 진행해보시면 어떨까 합니다! 🙇🏻‍♂️

언제든 추가적인 도움이 필요하다면 말씀해 주세요!

> 2️⃣ 공통부분 상속으로 구현했을 때 private 접근제어자 사용하지 못하는 문제
RegistrationView와 EditingView가 완벽히 겹치는 코드가 많아서 프로토콜과 상속을 통한 구현을 시도해봤습니다.
우선 protocol을 통해 구현하려고 했습니다. 저희 코드는 저장 프로퍼티를 많이 사용하고, 그 프로퍼티가 완벽히 겹칩니다. 때문에 protocol을 통해 구현하면 완벽히 같은 코드블럭도 똑같이 작성해야했습니다.(물론 코드 누락을 피할 수 있다는 장점이 생겼지만 중복코드가 여전히 존재하는 단점이...)
그래서 현재 ProductUpdator class를 통해 상속으로 구현했습니다. 하지만 super class의 모든 프로퍼티와 메서드의 접근제어자를 사용할 수 없어졌습니다. 결과적으로 은닉화/캡슐화가 잘된 것인가 라는 생각을 하게 됐습니다.
이 부분에서 저희가 사용한 상속이 올바른 방법인지 다른 방법이 있는지 궁금합니다.❤️‍🔥

음 혹시 요 부분에서 `super class의 모든 프로퍼티와 메서드의 접근제어자를 사용할 수 없어졌습니다.`라는 부분이 어떤 말씀이신지 조금 더 추가적인 설명을 부탁드려도 될까요? 🙏
어떤 상황인지 잘 그려지지가 않아서요..! 😓🙏

추가 설명 남겨주시면 이어서 답변드릴 수 있도록 하겠습니다!

> 3️⃣ 빈 문자열이 post 가능한 것에 대한 질문
위의 고민한 점에서(7️⃣ 빈 문자열을 get, post할 수 있다?) 언급했듯이 httpGet 했을때, price와 같은 Int 타입의 값에서 nil 값이 나옵니다!
>예상하기론
>
>1. 서버에서 빈 값을 nil로 바꾸어 저장했다.
> 2. 서버에서 Int 타입임에도 ""로 가지고 있다가 클라이언트로 넘어오면서 Int타입으로 받을 수 없어서 nil을 반환한다.
> 이렇게 두가지로 예상해봤습니다.
> 궁금한 점은 POST할 때 params의 price는 type이 number라고 기재되어있는데, 왜 빈배열("")을 POST 할 수 있는건지. 서버에서 어떤 방식으로 잘못된 정보를 취급하는 것인지 궁금합니다.🙇‍♀️

음 이 부분에 대해서는 제가 확실하게 답변을 드리기에는 어려울 것 같아요. 해당 서버에 대해서 알고있는게 하나도 없기 때문인데요..! 😇
postman으로 이것저것 API를 찔러보면서 Reverse Engineering을 시도해보거나, 조금 더 자세한 내용은 야곰에게 여쭈어 보는 방법으로 궁금증을 해결해 볼 수 있을것 같아요!

> 4️⃣ 이외에도 너무 어려운 것이 많아서 아직 질문조차 못 드리는 것이 많은 것 같아요. 잘 부탁드립니다 😭
음.. 시간 여유가 될지는 모르겠지만, 된다면 리뷰와는 별개로 프로젝트에 관해 이것저것 이야기해보는 시간을 가져보면 좋을것 같아요.
글로 옮기기 어려운 고민(질문)이나 어려움을 느낀 포인트들에 대해 일부 해소할 수 있을것 같아서요~!  🙌

막히시거나 궁금하신 부분이 있으면 언제든 말씀 부탁드립니다! 함께 고민해보아요 🙌
수고하셨습니다!

Commit:
♻️[refactor]: HttpProvider로 네이밍 변경 및 파일 이동

Commit:
♻️[refactor]: PageInformation으로 네이밍 변경

Commit:
♻️[refactor]: enum Currency 프로퍼티명 optionNumber로 변경

Commit:
♻️[refactor]: CodingKeys에 private 키워드 추가

JoSH0318 committed:
♻️[refactor]: extension String, integer 프로퍼티 반환값 변경

JoSH0318 committed:
🪛[chore]: 매직 넘버 제거

JoSH0318 committed:
♻️[refactor]: setupParams() 메서드, 조건 검사하는 메서드 분리 구현

JoSH0318 committed:
🐛[fix]: 잘못된 값이 label에 출력되는 버그 수정 및 컨벤션에 따른 개행

JoSH0318 commented:
@Monsteel  토니! 리팩토링 완료했습니다.
아마 키보드와 같은 경우는... 조금 더 시간이 필요할 것 같습니다.
1️⃣ 키보드와 같은 경우, 너무 사용중심으로 접근해서 저희처럼 특별한 케이스를 생각하지 못한것 같습니다. 좀더 근본적인 개념을 파악해야겠네요 😭
2️⃣ 공통부분 상속으로 구현했을 때 private 접근제어자 사용하지 못하는 문제
저희의 설명이 부족했습니다 😭
저희는 현재 두개의 view가 하나의 superView를 상속받고 있습니다. 두개의 view는 대부분의 프로퍼티와 메서드가 겹쳐서 superView에 구현을 했습니다. 그렇다보니 대부분의 프로퍼티와 메서드(superView에 구현된)에 private 접근제어자를 사용할 수 없게 됐는데요.
은닉화/캡슐화 시점에서 본다면 private 접근제어자가 없는 객체가 뭔가 불안정해보였습니다. 
상속을 한다면 superView에는 private 접근제어자를 사용할 수 없는데 이것이 안전한 방법일까요?

JoSH0318 committed:
🚚[move]: Endpoint, ImgaeCacheManager 파일 위치 변경

JoSH0318 committed:
🔥[del]: 쓸모없는 주석 제거

JoSH0318 committed:
✨[feat]: DetailVC, EditingVC final 키워드 추가

JoSH0318 committed:
✨[feat]: RegistrationVC 내 메서드 private 키워드 추가

JoSH0318 committed:
♻️[refactor]: MainVC 내부, enum Constants로 이름 변경

* private 키워드 추가

Commit:
🎨[style]: 팀 컨벤션에 따른 코드 구조 변경

JoSH0318 committed:
🎨[style]: 팀 컨벤션에 따른 코드 구조 변경2

JoSH0318 committed:
♻️[refactor]: NotificationCenter 관련 코드 제거 및 로직 변경

JoSH0318 committed:
🔥[del]: 개행 제거

JoSH0318 committed:
♻️[refactor]: 각 VC에 키보드 구현

*UITextViewDelegate 채택을 통해
textViewDidBeginEditing, textViewDidEndEditing 메서드를
커스텀하여 동적 키보드 구현

JoSH0318 committed:
🔥[del]: 쓸모없는 코드 제거

Monsteel commented:
@JoSH0318 
답변 감사합니다! :)

> 1️⃣ 키보드와 같은 경우, 너무 사용중심으로 접근해서 저희처럼 특별한 케이스를 생각하지 못한것 같습니다. 좀더 근본적인 개념을 파악해야겠네요 😭

좋습니다! 조금 더 시간을 가지고 고민해보셔도 좋아요 👍

> 2️⃣ 공통부분 상속으로 구현했을 때 private 접근제어자 사용하지 못하는 문제
저희의 설명이 부족했습니다 😭
저희는 현재 두개의 view가 하나의 superView를 상속받고 있습니다. 두개의 view는 대부분의 프로퍼티와 메서드가 겹쳐서 superView에 구현을 했습니다. 그렇다보니 대부분의 프로퍼티와 메서드(superView에 구현된)에 private 접근제어자를 사용할 수 없게 됐는데요.
은닉화/캡슐화 시점에서 본다면 private 접근제어자가 없는 객체가 뭔가 불안정해보였습니다.
상속을 한다면 superView에는 private 접근제어자를 사용할 수 없는데 이것이 안전한 방법일까요?

아 저도 뭔가 이런 상황을 고민해보았던 적이 있었는데요, 다른 언어에서는 protected와 같은 개념을 활용하여 일부 해결할 수 있었을텐데,
Swift에서는 protected개념이 없어 고민되는건 사실이지요. 
(cf. [Why doesn’t Swift have something like protected? 라는 질문으로 시작하는 apple dev blog 글](https://developer.apple.com/swift/blog/?id=11))

private 선언에 대해 강박을 버리고, 외부에서 접근은 가능하되 수정은 안되도록(상수) 처리를 해두는 선에서 저는 고민을 마무리 지었었는데요,

꼭 private 선언이 필요하다면, protocol로 구현해야 할 value(상수 혹은 변수)를 정의하고,
공통되게 사용하는 인스턴스를 만드는 외부 객체(Factory) 를 만들어서 주입하는 방법이 있을 수 있을것 같긴 합니다! 🙋🏻‍♂️

JoSH0318 committed:
♻️[refactor]: 키보드 editing 때, textView 높이 변경하는 로직 추가

* textViewDidBeginEditing() 내부에
textView.contentInset.bottom = raisingHeight 할당하여
키보드 editing 일때, contentInset.bottom 변경되도록 수정

Commit:
✨[feat]: 이미지 fileName을 고유한 값인 날짜, 시간으로 부여하는 로직 구현

JoSH0318 commented:
@Monsteel 토니! 남겨주신 답변 너무너무 감사합니다.
문제가 됐고, 고민했던 것들의 문제를 해결한 내용입니다.😀

1️⃣ 키보드 문제
키보드가 올라올 때, NotificationCenter를 통해서 키보드가 올라오면 view frame을 바꿔주는 형태였습니다.
저희의 경우 textField(이름, 가격, 할인가격, 재고수량)는 view frame은 변하지 않기를 원했고, 
NotificationCenter를 사용하면 키보드가 올라올 때마다 동작하게 되어서 저희가 원하는 동작을 할 수가 없더라구요!
그래서! NotificationCenter를 없앴습니다!

대신, UITextViewDelegate 프로토콜의 delegate 메서드로도 충분히 구현할 수 있었습니다.🪄
textViewDidBeginEditing(), textViewDidEndEditing() 메서드를 이용해서 editing이 시작, 끝나는 시점을 알 수 있었고, 
해당 시점에 원하는 동작(Frame 변경)을 해주었습니다.
추가적으로 text가 길어져서 라인이 길어지는 경우 contentInset bottom을 변경해서 키보드에 가리지 않도록 했습니다.
그리고 UIPanGestureRecognizer를 이용해서 키보드를 EndEditing하는 제스처를 주었습니다.
![](https://i.imgur.com/plgTRJO.gif)

2️⃣ Image FileName
토니의 코멘트가 너무 좋은 답변이 됐습니다! 🙇‍♂️
Date와 DateFormatter를 이용해서 ImageFile 인스턴스를 생성하는 시간에 따라서 fileName을 정하도록 했습니다.

3️⃣ 상속과 private
저희가 private에 대한 강박이 생겼던 것 같습니다!
특히 저희의 경우 class의 프로퍼티들이 대부분 값이 아닌 UI요소들입니다. 
따라서 그 값을 조회한다고 해서 크게 문제가 되진 않을 것 같습니다.
오히려 말씀하신대로 let으로 값을 변하지 못하게 하는 것만으로 충분할 것 같다는 생각을 했습니다.

4️⃣ pagination
이전에 문제가 되었던 pagination(페이징 기능)을 수정했습니다.
이전에는 indexPath.row를 기준으로 페이징을 했었는데 
토니의 조언대로 contentOffset을 기준으로 페이징되도록 바꿨습니다.
확실히 저희가 컨트롤할 수 없는 영역인 indexPath.row를 사용하는 것은 불안정하다고 느꼈습니다.

Monsteel reviewed (approved):
밈 캐시 수고많으셨습니다~! 이어서 Step2 진행해 주시면 감사하겠습니다~! 🙌
